### ISSUE TITLE
Choppy Audio -- Only in AssistantMk2

### ISSUE THREAD

**Issue by ItzAlwaysDNS** [eouia/MMM-AssistantMk2#77]
Hello,

I just finished installing AssistantMk2 as well as Hotword. They both work great! However whenever I ask the mirror a request through the Hotword module, the audio playback (HDMI via Monitor speakers) sounds very choppy. After doing some research, it appears this could be due to high CPU usage, however I don't think this is the case. 

If I play the following command without the Hotword module, the audio is perfectly fine and not choppy:
speaker-test -t wav

*CPU usage ~30%
*Possibly something to do with mpg321? I installed this as per the installation guide. I'm not sure if the above command uses the mpg321. Can I test this by removing the program? If so, how could I go about doing that?

**Comment by eouia** [comment]
Test it like this.
Before playback of Response voice is completely ended, Stop the MM.
Then go to `tmp` directory of this module. There could be some XXXXXXX.mp3. play it with your mpg321 or anything.

Via HDMI there could be some time lag to play sound. Try other play program. These program is available. (Some configuration modifying is needed.)
https://github.com/eouia/MMM-AssistantMk2/blob/master/INSTALL.md#2-speaker-for-playing

**Comment by ItzAlwaysDNS** [comment]
> Test it like this.
> Before playback of Response voice is completely ended, Stop the MM.
> Then go to `tmp` directory of this module. There could be some XXXXXXX.mp3. play it with your mpg321 or anything.
> 
> Via HDMI there could be some time lag to play sound. Try other play program. These programs are available. (Some configuration modifying is needed.)
> https://github.com/eouia/MMM-AssistantMk2/blob/master/INSTALL.md#2-speaker-for-playing

Whenever I try playing it back with mpg321, I'm getting "Invalid Argument"
~/MagicMirror/modules/MMM-AssistantMk2/tmp $ mpg321 test.mp3

Now that you mention that, I'm guessing that the Assistant is playing the mp3 files from Hotword through aplay, which is unsupported for mp3 files. How can I ensure/verify that the Assistant is playing back through mpg321 not aplay?

Thank you for your fast response!

**Comment by ItzAlwaysDNS** [comment]
Below is my config file that I'm working with. Do I need to add anything under the AssistantMk2 module to specify mpg321?

/* Magic Mirror Config Sample
 *
 * By Michael Teeuw http://michaelteeuw.nl
 * MIT Licensed.
 *
 * For more information how you can configurate this file
 * See https://github.com/MichMich/MagicMirror#configuration
 *
 */

var config = {
	address: "localhost", // Address to listen on, can be:
	                      // - "localhost", "127.0.0.1", "::1" to listen on loopback interface
	                      // - another specific IPv4/6 to listen on a specific interface
	                      // - "", "0.0.0.0", "::" to listen on any interface
	                      // Default, when address config is left out, is "localhost"
	port: 8080,
	ipWhitelist: ["127.0.0.1", "::ffff:127.0.0.1", "::1"], // Set [] to allow all IP addresses
	                                                       // or add a specific IPv4 of 192.168.1.5 :
	                                                       // ["127.0.0.1", "::ffff:127.0.0.1", "::1", "::ffff:192.168.1.5"],
	                                                       // or IPv4 range of 192.168.3.0 --> 192.168.3.15 use CIDR format :
	                                                       // ["127.0.0.1", "::ffff:127.0.0.1", "::1", "::ffff:192.168.3.0/28"],

	language: "en",
	timeFormat: 24,
	units: "metric",

	modules: [
		{
			module: "alert",
		},
		{
			module: "updatenotification",
			position: "top_bar"
		},
		{
			module: "clock",
			position: "top_left"
		},
		{
			module: "calendar",
			header: "US Holidays",
			position: "top_left",
			config: {
				calendars: [
					{
						symbol: "calendar-check-o ",
						url: "webcal://www.calendarlabs.com/templates/ical/US-Holidays.ics"
					}
				]
			}
		},
		{
			module: "compliments",
			position: "lower_third"
		},
		{
			module: "currentweather",
			position: "top_right",
			config: {
				location: "New York",
				locationID: "",  //ID from http://bulk.openweathermap.org/sample/; unzip the gz file and find your city
				appid: "YOUR_OPENWEATHER_API_KEY"
			}
		},
		{
			module: "weatherforecast",
			position: "top_right",
			header: "Weather Forecast",
			config: {
				location: "New York",
				locationID: "5128581",  //ID from http://www.openweathermap.org/help/city_list.txt
				appid: "YOUR_OPENWEATHER_API_KEY"
			}
		},
		{
			module: "newsfeed",
			position: "bottom_bar",
			config: {
				feeds: [
					{
						title: "New York Times",
						url: "http://www.nytimes.com/services/xml/rss/nyt/HomePage.xml"
					}
				],
				showSourceTitle: true,
				showPublishDate: true
			}
		},
		{
		  module: "MMM-AssistantMk2",
		  position: "top_right",
		  
       		  config: {
		  	record: {
		  		recordProgram : "arecord",  
		  		device        : "plughw:1",
		  	   },
		
			   notifications: {
				ASSISTANT_ACTIVATED: "HOTWORD_PAUSE",
				ASSISTANT_DEACTIVATED: "HOTWORD_RESUME",
      },

		  }
		},
		{
		  module: "MMM-Hotword",
		  config: {
		    record: {
		      recordProgram : "arecord",  
		      device        : "plughw:1"
		    },
		    autostart:true,
		    onDetected: {
		      notification: (payload) => {
		        return "ASSISTANT_ACTIVATE"
		      },
		      payload: (payload) => {
		        return {
		          profile: payload.hotword
		        }
		      }
		    },
		  },
		},


	]

};

/*************** DO NOT EDIT THE LINE BELOW ***************/
if (typeof module !== "undefined") {module.exports = config;}

**Comment by eouia** [comment]
mpg321 is default player program. If you want to use other, adjust configuration
```
 play: { // Full version
      encodingOut: "MP3", //'MP3' or 'WAV' is available, but you might not need to modify this.
      sampleRateOut: 24000,
      playProgram: "mpg321", //Your prefer sound play program. By example, if you are running this on OSX, `afplay` could be available.
      playOption: [],
      // If you need additional options to use playProgram, describe here. (except filename)
      // e.g: ["-d", "", "-t", "100"]
    },
```

**Comment by eouia** [comment]
If you want to use `aplay` , I don't recommend. (Too many configuration modifications are needed)
Install other programs for mp3 and try them.
https://github.com/eouia/MMM-AssistantMk2/blob/master/INSTALL.md#2-speaker-for-playing

**Comment by ItzAlwaysDNS** [comment]
> mpg321 is default player program. If you want to use other, adjust configuration
> 
> ```
>  play: { // Full version
>       encodingOut: "MP3", //'MP3' or 'WAV' is available, but you might not need to modify this.
>       sampleRateOut: 24000,
>       playProgram: "mpg321", //Your prefer sound play program. By example, if you are running this on OSX, `afplay` could be available.
>       playOption: [],
>       // If you need additional options to use playProgram, describe here. (except filename)
>       // e.g: ["-d", "", "-t", "100"]
>     },
> ```

This is what I was looking for, thanks. It appears my config is set to use mpg321.

Its odd that .wav files played through the arecord are perfectly fine, yet mp3 files played back through mpg321 are choppy. However, it makes sense if the mp3 file created by the record function is choppy if it is playing back through a non-compatible player (like aplay).

When I'm trying to playback the mp3 file in the tmp directory, I keep getting an argument error. Is the syntax not just:

"mpg321 /tmp/test.mp3" (without the quotes and in the directory where the mp3 resides)?

**Comment by ItzAlwaysDNS** [comment]
> ```
> cd ~/MagicMirror/modules/MMM-AssistantMk2/tmp/
> mpg321 test.mp3
> ```
> will work.
> Anyway, HDMI output is not so good. Someone told me he has a similar issue of output quality through HDMI. He Had changed to analog speaker or BT speaker, that issue was fixed.

Thanks for the verification. Its odd because before I added the hotword module, the audio was fine. Adding the sample configuration with the "briefing today" for the MMM-AssistantMk2  in ~/MagicMirror/config/config.js works perfectly. Only after adding the hotword into the configuration did the audio get choppy. Will post back later tonight with logs and config.

**Comment by eouia** [comment]
It could be performance issue. MMM-Hotword uses `snowboy` as detection engine. It is running on background and it could make whole performance low. If you have many modules or many background process on your RPI, this could be the the reason.

**Comment by ItzAlwaysDNS** [comment]
> It could be performance issue. MMM-Hotword uses `snowboy` as detection engine. It is running on background and it could make whole performance low. If you have many modules or many background process on your RPI, this could be the the reason.

UPDATE:

Issue resolved. I ended up removing mpg321 (sudo apt-get remove mpg321) and my choppy audio seems to have gone away. I tried reinstalling it, but the choppy audio comes right back. For now, I'm keeping it uninstalled. Magic Mirror and the AssistantMk2 seem to be working fine, despite making 0 configuration changes after removal. 

Thanks for all the help eouia!
